sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,634 @@
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.util.NamedList;
+
+/**
+ * Transforms for reading and writing data from/to Solr.
+ *
+ * <h3>Reading from Solr</h3>
+ *
+ * <p>{@link SolrIO#read SolrIO.read()} returns a bounded
+ * {@link PCollection PCollection&lt;SolrDocument&gt;} representing Solr documents.
+ *
+ * <p>To configure the {@link SolrIO#read}, you have to provide a connection configuration
+ * containing the Zookeeper address of the Solr cluster. The following example
+ * illustrates options for configuring the source:
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(SolrIO.read().withConnectionConfiguration(
+ *    SolrIO.ConnectionConfiguration.create("127.0.0.1:9983", "my-collection")
+ * )
+ *
+ * }</pre>
+ *
+ * <p>You can specify a query on the {@code read()} using {@code withQuery()}.
+ *
+ * <h3>Writing to Solr</h3>
+ *
+ * <p>To write documents to Solr, use
+ * {@link SolrIO#write SolrIO.write()}, which writes Solr documents from a
+ * {@link PCollection PCollection&lt;SolrInputDocument&gt;} (which can be bounded or unbounded).
+ *
+ * <p>To configure {@link SolrIO#write SolrIO.write()}, similar to the read, you
+ * have to provide a connection configuration. For instance:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...)
+ *    .apply(SolrIO.write().withConnectionConfiguration(
+ *       SolrIO.ConnectionConfiguration.create("127.0.0.1:9983", "my-collection")
+ *    )
+ *
+ * }</pre>
+ *
+ * <p>Optionally, you can provide {@code withBatchSize()}
+ * to specify the size of the write batch in number of documents.
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class SolrIO {
+
+  public static Read read() {
+    return new AutoValue_SolrIO_Read.Builder().setBatchSize(100L).setQuery("*:*").build();
+  }
+
+  public static Write write() {
+    return new AutoValue_SolrIO_Write.Builder().setMaxBatchSize(1000L).build();
+  }
+
+  private SolrIO() {
+  }
+
+  /** A POJO describing a connection configuration to Solr. */
+  @AutoValue
+  public abstract static class ConnectionConfiguration implements Serializable {
+
+    abstract String getZkHost();
+    abstract String getCollection();
+    @Nullable abstract String getUsername();
+    @Nullable abstract String getPassword();
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setZkHost(String zkHost);
+      abstract Builder setCollection(String collection);
+      abstract Builder setUsername(String username);
+      abstract Builder setPassword(String password);
+      abstract ConnectionConfiguration build();
+    }
+
+    /**
+     * Creates a new Solr connection configuration.
+     *
+     * @param zkHost host of zookeeper
+     * @param collection the collection toward which the requests will be issued
+     * @return the connection configuration object
+     * @throws IOException when it fails to connect to Solr
```

Comment:
This method can't throw an IOException, it only builds a simple Java object but doesn't actually connect